lib/pbench/test/unit/server/test_datasets_metadata.py

```diff
@@ -0,0 +1,178 @@
+                "name": "drb",
+                "metadata": ["seen", "saved", "access", "user"],
+            },
+        )
+        assert response.status_code == HTTPStatus.OK
+        assert response.json == {
+            "saved": False,
+            "seen": True,
+            "access": "private",
+            "user": {"contact": "me@example.com"},
+        }
+
+    def test_put_missing_json_object(self, client, server_config, pbench_token):
+        """
+        Test behavior when no JSON payload is given
+        """
+        response = client.put(f"{server_config.rest_uri}/datasets/metadata")
+        assert response.status_code == HTTPStatus.BAD_REQUEST
+        assert response.json.get("message") == "Invalid request payload"
+
+    def test_put_missing_keys(self, client, server_config, user_ok, pbench_token):
+        """
+        Test behavior when JSON payload does not contain all required keys.
+
+        Note that Pbench will silently ignore any additional keys that are
+        specified but not required.
+
+        TODO: This is mostly copied from commons.py
+        """
+        classobject = DatasetsMetadata(client.config, client.logger)
+
+        def missing_key_helper(keys):
+            response = client.post(
+                f"{server_config.rest_uri}/datasets/metadata", json=keys,
+            )
+            assert response.status_code == HTTPStatus.BAD_REQUEST
+            missing = [k for k in required_keys if k not in keys]
+            assert (
+                response.json.get("message")
+                == f"Missing required parameters: {','.join(missing)}"
+            )
+
+        parameter_items = classobject.schema.parameters.items()
+
+        required_keys = [
+            key for key, parameter in parameter_items if parameter.required
+        ]
+
+        all_combinations = []
+        for r in range(1, len(parameter_items) + 1):
+            for item in itertools.combinations(parameter_items, r):
+                tmp_req_keys = [key for key, parameter in item if parameter.required]
+                if tmp_req_keys != required_keys:
+                    all_combinations.append(item)
+
+        for items in all_combinations:
+            keys = {}
+            for key, parameter in items:
+                if parameter.type is ParamType.ACCESS:
+                    keys[key] = "public"
+                elif parameter.type is ParamType.DATE:
+                    keys[key] = "2020"
+                elif parameter.type is ParamType.LIST:
+                    keys[key] = []
+                elif parameter.type is ParamType.KEYWORD:
+                    keys[key] = parameter.keywords[0]
+                else:
+                    keys[key] = "foobar"
+
+            missing_key_helper(keys)
+
+        # Test in case all of the required keys are missing and some
+        # random non-existent key is present in the payload
+        if required_keys:
+            missing_key_helper({"notakey": None})
```

Comment:
If I might suggest...I would move this inside the loop -- call `missing_key_helper()` once with the generated value of `keys` and then call it again with `{"notakey": None}` added to `keys` (I don't think the second call would need to be conditional).

(I would also be inclined to start the loop at line 87 with zero instead of one; I believe Nikhil said that calling his equivalent of `missing_key_helper()` didn't work with an empty list, but we could skip the first of the two calls in that case and only do the second, which would cover the case that lines 111-112 currently cover.)

Reply:
This was just copied from Nikhil's test; and I'm more inclined to keep them consistent (in the hopes they'll later be refactored out a layer), and less inclined at this point to modify both to keep them consistent.